radix-engine/src/state_manager/staging.rs

```diff
@@ -0,0 +1,216 @@
+use sbor::rust::collections::*;
+use sbor::rust::vec::Vec;
+use scrypto::buffer::{scrypto_decode, scrypto_encode};
+
+use crate::ledger::*;
+
+/// Nodes form an acyclic graph towards the parent
+struct StagedExecutionStoreNode {
+    parent_id: u64,
+    locked: bool,
+    spaces: BTreeMap<Vec<u8>, OutputId>,
+    outputs: BTreeMap<Vec<u8>, Output>,
+}
+
+impl StagedExecutionStoreNode {
+    fn new(parent_id: u64) -> Self {
+        StagedExecutionStoreNode {
+            parent_id,
+            locked: false,
+            spaces: BTreeMap::new(),
+            outputs: BTreeMap::new(),
+        }
+    }
+}
+
+/// Structure which manages the acyclic graph
+pub struct StagedExecutionStores<'s, S: ReadableSubstateStore + WriteableSubstateStore> {
+    parent: &'s mut S,
+    nodes: HashMap<u64, StagedExecutionStoreNode>,
```

Comment:
Can we have a proper tree data structure instead of `HashMap`?

Reply:
We need an index into every node as well as a way to remove nodes in the middle of the tree. Is there a Rust data structure for that?

Reply:
That's exactly one of the reasons we need tree structures. 

It can be as simple as:
```
Node {
   node_state: BlaBla,
   children: Vec<Node>
}
```

Reply:
Yup like mentioned, this has repercussions in terms of how do we index to each node. Like git/PMT we need a map from id => state.